### PR TITLE
[BEAM-79] Add SideInput support for GearpumpRunner

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by manuzhang** [comment]
R: @kennknowles 

travis failed with 
```
testJodaTimeAndJava8TimeConversion(org.apache.beam.runners.gearpump.translators.utils.TranslatorUtilsTest)  Time elapsed: 0.006 sec  <<< ERROR!
java.lang.ArithmeticException: long overflow
```
which we found happened with older JDK version (see [comments here](https://github.com/apache/incubator-gearpump/pull/158#issuecomment-282192920))

**Comment by kennknowles** [comment]
The JDK ticket implies it is fixed only in JDK 9? That would be quite a problem.

Do you need those min/max values? We actually only use a smaller subset for millis, from `BoundedWindow.TIMESTAMP_MIN_VALUE` to `BoundedWindow.TIMESTAMP_MAX_VALUE`. This is so if a runner chooses to use microseconds internally, there will not be an overflow.

**Comment by manuzhang** [comment]
@kennknowles It's fixed in newer versions of JDK 8 (e.g. jdk1.8.0_121) and verified in Gearpump's [integration test](https://github.com/apache/incubator-gearpump/tree/master/integrationtest/docker/java). I don't have a rational behind using min/max values. It's just the boundary which shouldn't cause overflow for Gearpump (where millis is used). Using a smaller subset looks a bit strange there.

**Comment by manuzhang** [comment]
Travis [suggests](https://docs.travis-ci.com/user/languages/java/) we add these lines to update to latest oracleJDK 8

```
sudo: false
addons:
  apt:
    packages:
      - oracle-java8-installer
```

WDYT?

**Comment by kennknowles** [comment]
In fact, I think we should retire Travis-CI very soon except for basic smoke tests of environments that we cannot exercise through Jenkins. We have hit its timeout limit and we do not really have many options to fix that.

I believe the JDK7 vs JDK8 will be tested well via Jenkins, while just a basic (very quick) test on OSX will remain on Travis-CI. I don't know a timeline on getting this rolled out. Please, @davorbonaci @jasonkuster, note this issue related to differing versions of JDK8 w.r.t. this JDK bug.

For the feature branch, doing that sort of configuration is OK, but when `master` changes this branch should follow it.

**Comment by kennknowles** [comment]
(I won't block merge on Travis or this issue)

**Comment by jasonkuster** [comment]
#2204 is @markflyhigh 's initial work to get the cross-JDK testing working and it looks like it's progressing okay, so I'd say it looks like timeline is relatively soon.

**Comment by manuzhang** [comment]
@kennknowles @jasonkuster Thanks for the inputs.
Could we leave it as it is now and sync with master for travis changes before merging ?

**Comment by kennknowles** [comment]
Yes, that is fine. Travis is not our primary signal and certainly doesn't need to be kept green on this branch.

**Comment by kennknowles** [comment]
I haven't had time yet to thoroughly review this, that is why I have so few comments. I will get to it soon, though.

File: runners/gearpump/pom.xml [hunk 1]
```diff
@@ -75,23 +75,9 @@
...
-                      org.apache.beam.sdk.transforms.ApproximateUniqueTest,
-                      org.apache.beam.sdk.transforms.CombineTest,
-                      org.apache.beam.sdk.transforms.CombineFnsTest,
-                      org.apache.beam.sdk.transforms.CountTest,
-                      org.apache.beam.sdk.transforms.FlattenTest,
-                      org.apache.beam.sdk.transforms.ParDoTest,
-                      org.apache.beam.sdk.transforms.SampleTest,
-                      org.apache.beam.sdk.transforms.ViewTest,
-                      org.apache.beam.sdk.transforms.join.CoGroupByKeyTest
+                      org.apache.beam.sdk.transforms.ParDoTest
```
Comment (manuzhang) [inline runners/gearpump/pom.xml]: The last piece is `TimerInternals` and `StateInternals`. Not sure there is such a tag.

Comment (kennknowles) [inline runners/gearpump/pom.xml]: Can you remove it and run them? I may have suggestions that we can add.

For user-facing state and timers there is `UsesStateInParDo` and `UsesTimersInParDo`.

Comment (manuzhang) [inline runners/gearpump/pom.xml]: Yes, it's working

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpRunner.java [hunk 2]
```diff
@@ -69,15 +98,23 @@ public static GearpumpRunner fromOptions(PipelineOptions options) {
 
   public <OutputT extends POutput, InputT extends PInput> OutputT apply(
       PTransform<InputT, OutputT> transform, InputT input) {
-    if (Flatten.FlattenPCollectionList.class.equals(transform.getClass())
+    if (overrides.containsKey(transform.getClass())) {
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpRunner.java]: The next time you sync from master, this method will go away and you'll need to do all of this via `runners-core-construction` utilities. So let's definitely do this PR first so you don't have too much stuff going on at once.

File: runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpRunner.java [hunk 3]
```diff
@@ -131,4 +168,320 @@ private Config registerSerializers(Config config, Map<String, String> userSerial
...
+      @SuppressWarnings({"rawtypes", "unchecked"})
+      KvCoder<K, V> inputCoder = (KvCoder) input.getCoder();
+      try {
+        inputCoder.getKeyCoder().verifyDeterministic();
+      } catch (Coder.NonDeterministicException e) {
+        // throw new RuntimeException(e);
+      }
+
+      return input
+          .apply(Combine.globally(new Concatenate<KV<K, V>>()).withoutDefaults())
```
Comment (kennknowles) [inline runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/GearpumpRunner.java]: I believe a lot of this is now actually implemented as part of the default expansion. So this code my go away after a sync from master.

**Pull request merged**